### DISCUSSION TITLE
typing a type-or-object function

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#4412] [Q&A]
Hypothetically, I have this function, and I can't change it's implementation but it's used in my code base and I want the strongest typing available:

```
def copy_or_instance(x):
  if isinstance(x, type):
    return x()
  else:
    return x.copy()
```

Is this the best way to annotate such a beast? It took me a while to realize I needed to use more than one typevar here.

```
@t.overload
def f(x: type[T]) -> T:
    ...


@t.overload
def f(x: S) -> S:
    ...


def f(x: S | type[T]) -> S | T:
    if isinstance(x, type):
        return x()
    else:
        return x
```

**Comment by erictraut** [comment]
Yes, that's how I'd recommend annotating it.
